packages/react/src/components/UIShell/Header.js

```diff
@@ -8,16 +8,41 @@
 import { settings } from 'carbon-components';
 import cx from 'classnames';
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, { useState } from 'react';
 import { AriaLabelPropType } from '../../prop-types/AriaPropTypes';
 
 const { prefix } = settings;
 
-const Header = ({ className: customClassName, children, ...rest }) => {
+export const HandleMenuButtonClickContext = React.createContext();
+
+const Header = ({
+  className: customClassName,
+  children,
+  isSideNavExpanded,
+  ...rest
+}) => {
   const className = cx(`${prefix}--header`, customClassName);
+
+  const [isSideNavExpandedState, setIsSideNavExpandedState] = useState(
+    isSideNavExpanded
+  );
+
+  const handleHeaderMenuButtonClick = () => {
+    const newSideNavExpandedState = !isSideNavExpandedState;
+    setIsSideNavExpandedState(newSideNavExpandedState);
+    return newSideNavExpandedState;
+  };
+
+  const childrenProps = {
+    function: handleHeaderMenuButtonClick,
+    state: isSideNavExpandedState,
+  };
+
   return (
     <header {...rest} className={className} role="banner">
-      {children}
+      <HandleMenuButtonClickContext.Provider value={childrenProps}>
```

Comment:
Is context usage here is for [the same problem discussed in #2786](https://github.com/carbon-design-system/carbon/pull/2786#issuecomment-494174254)? If so, my recommendation is creating a "state manager" like:

```javascript
const HeaderContainer = ({ isSideNavExpanded, onClickSideNavExpando, render: Children }) => {
  const [isSideNavExpandedState, setIsSideNavExpandedState] = useState(isSideNavExpanded);
  const handleClickSideNavExpando = useCallback(() => {
    if (typeof isSideNavExpanded === 'undefined') {
      const newIsSideNavExpandedState = !isSideNavExpandedState;
      setIsSideNavExpandedState(newIsSideNavExpandedState, () => {
        if (onClickSideNavExpando) {
          onClickSideNavExpando({ isSideNavExpanded: newIsSideNavExpandedState });
        }
      });
    } else if (onClickSideNavExpando) {
      onClickSideNavExpando({ isSideNavExpanded: !isSideNavExpanded });
    }
  }, [isSideNavExpandedState, setIsSideNavExpandedState, isSideNavExpanded, onClickSideNavExpando]);
  return (
    <Children
      isSideNavExpanded={typeof isSideNavExpanded !== 'undefined' ? isSideNavExpanded : isSideNavExpandedState}
      onClickSideNavExpando={handleClickSideNavExpando}
    />
  );
};
```

Which can be used like:

```javascript
<HeaderContainer render={({ isSideNavExpanded, onClickSideNavExpando }) => (
  <>
    <Header aria-label="IBM Platform Name">
      <SkipToContent />
      <HeaderMenuButton
        aria-label="Open menu"
        onClick={onClickSideNavExpando}
      />
      ...
    </Header>
    <SideNav isSideRail aria-label="Side navigation" isExpanded={isSideNavExpanded}>
        ...
    </SideNav>
    <StoryContent />
  </>
)}>
```

Reply:
Yes, this is the same issue as that comment. State manager like ☝️ may make maintainability a bit easier in the future.